kafka-connect-s3/src/test/java/io/confluent/connect/s3/integration/BaseConnectorIT.java

```diff
@@ -0,0 +1,153 @@
+   *
+   * @param connectorName the connector
+   * @param numTasks the minimum number of tasks
+   * @return true if the connector and tasks are in RUNNING state; false otherwise
+   */
+  protected Optional<Boolean> assertConnectorAndTasksRunning(String connectorName, int numTasks) {
+    try {
+      ConnectorStateInfo info = connect.connectorStatus(connectorName);
+      boolean result = info != null
+                       && info.tasks().size() >= numTasks
+                       && info.connector().state().equals(AbstractStatus.State.RUNNING.toString())
+                       && info.tasks().stream().allMatch(s -> s.state().equals(AbstractStatus.State.RUNNING.toString()));
+      return Optional.of(result);
+    } catch (Exception e) {
+      log.error("Could not check connector state info.", e);
+      return Optional.empty();
+    }
+  }
+
+  protected void waitForConnectorToCompleteSendingRecords(long noOfRecordsProduced, int flushSize, String bucketname) throws InterruptedException {
+    TestUtils.waitForCondition(
+      () -> assertSuccess(noOfRecordsProduced, flushSize, bucketname).orElse(false),
+      CONNECTOR_STARTUP_DURATION_MS,
+      "Connector could not send all records in time."
+    );
+  }
+
+  protected Optional<Boolean> assertSuccess(long noOfRecordsProduced, int flushSize, String bucketname) {
+    try {
+    int noOfObjectsInS3 = getNoOfObjectsInS3(bucketname);
+    boolean result = noOfObjectsInS3 == Math.ceil(noOfRecordsProduced/flushSize);
+    return Optional.of(result);
+  } catch (Exception e) {
+    log.error("Could not check S3 state", e);
+    return Optional.empty();
+  }
+  }
+
+  protected int getNoOfObjectsInS3(String bucketName) {
+    ListObjectsV2Request req = new ListObjectsV2Request().withBucketName(bucketName).withMaxKeys(100);
+    ListObjectsV2Result result;
+    List<S3Object> records = new ArrayList<>();
+
+    do {
+      result = s3RootClient.listObjectsV2(req);
+
+      for (S3ObjectSummary objectSummary : result.getObjectSummaries()) {
+        if(objectSummary.getSize() > 0) {
+          records.add(s3RootClient.getObject(bucketName, objectSummary.getKey()));
+        }
+      }
+      // If there are more than maxKeys keys in the bucket, get a continuation token
+      // and list the next objects.
+      String token = result.getNextContinuationToken();
+      req.setContinuationToken(token);
+    } while (result.isTruncated());
+    return records.size();
+  }
+
+  protected int waitForFetchingStorageObjectsInS3(String bucketName) throws InterruptedException {
+    return waitForFetchingStorageObjectsInS3(bucketName, CONSUME_MAX_DURATION_MS);
+  }
+
+  protected int waitForFetchingStorageObjectsInS3(String bucketName, long maxWaitMs) throws InterruptedException {
+    long startTime = System.currentTimeMillis();
+    while(System.currentTimeMillis() - startTime < maxWaitMs) {
+      Thread.sleep(Math.min(maxWaitMs, 100L));
```

Comment:
Instead of waiting get the objects count here and if it matches the expected value then break.